src/guard.c

```diff
@@ -0,0 +1,63 @@
+#include "jemalloc/internal/jemalloc_preamble.h"
+#include "jemalloc/internal/jemalloc_internal_includes.h"
+
+#include "jemalloc/internal/assert.h"
+#include "jemalloc/internal/guard.h"
+#include "jemalloc/internal/pages.h"
+#include "jemalloc/internal/tsd.h"
+
+/* The sanitizer options. */
+size_t opt_san_guard_large_nextents = SAN_GUARD_LARGE_N_EXTENTS;
+size_t opt_san_guard_small_nextents = SAN_GUARD_SMALL_N_EXTENTS;
+
+void
+guard_pages(tsdn_t *tsdn, edata_t *edata, emap_t *emap) {
+	emap_deregister_boundary(tsdn, emap, edata);
+
+	size_t size_with_guards = edata_size_get(edata);
+	size_t usize = size_with_guards - PAGE_GUARDS_SIZE;
+
+	uintptr_t guard1 = (uintptr_t)edata_base_get(edata);
+	uintptr_t addr = guard1 + PAGE;
+	uintptr_t guard2 = addr + usize;
+
+	assert(edata_state_get(edata) == extent_state_active);
+	pages_mark_guards((void *)guard1, (void *)guard2);
```

Comment:
These should get indirected through the ehooks, a la ehooks_zero (and, like dalloc_will_fail, it's probably worth having an ehooks_guard_will_fail to handle the non-default case).

Reply:
You mean introducing a hook for adding guards? That's a good point; but I think that should wait until hook V2? Cause right now we probably don't want to modify the hook struct (and the guard logic is only enabled for default hooks right now).

Reply:
Right; I don't mean modify extent_hook_t; just indirecting through the ehooks interface (in the same way that ehooks_zero isn't really an extent hook; it's just an interface wrapper). The idea being that the ehooks module can evolve into the extent hooks v2 interface.

E.g. the checks against "are these default ehooks" and whatnot can be hidden behind that boundary.

Reply:
Makes sense. Will update.